### ISSUE TITLE
Unhandled Java exception: org.jruby.RubyContinuation$Continuation

### ISSUE THREAD

**Issue by Confusion** [jruby/jruby#6507]
When running the following script
```
require 'parser/current'

code = "def main; CONSTANT = nil; end"

parser = Parser::CurrentRuby.new
buffer = Parser::Source::Buffer.new('code', source: code)
parser.parse(buffer)
```
as `bin/jruby repro.rb`, the script fails with:
```
Unhandled Java exception: org.jruby.RubyContinuation$Continuation
org.jruby.RubyContinuation$Continuation: null
```
The expected behavior for this way of using the parser gem is that the script silently ends. There are ways to get it to report parse failures (`parser.diagnostics.consumer = ->(diag) {puts diag.render}`), but those don't otherwise change the behavior. 

This is using the whitequark/parser gem, 3.0.0.0. The issue was originally found using parser 2.7.2.0 on JRuby 9.2.13.0.

## Further context
This issue is related to a combination of parser, racc and JRuby. Similar code is otherwise handled as expected:
```
$ bin/jruby -e 'def main; CONSTANT = nil; end'
SyntaxError: -e:1: dynamic constant assignment
```

With `Parser::CurrentRuby.parse(code)` instead of the Parser invocation in the reproduction:
```
$ bin/jruby repro.rb 
(string):2:3: error: dynamic constant assignment
(string):2:   CONSTANT = nil
(string):2:   ^~~~~~~~      
Parser::SyntaxError: dynamic constant assignment
           process at jruby/lib/ruby/gems/shared/gems/parser-3.0.0.0/lib/parser/diagnostic/engine.rb:72
        diagnostic at jruby/lib/ruby/gems/shared/gems/parser-3.0.0.0/lib/parser/builders/default.rb:2189
        assignable at jruby/lib/ruby/gems/shared/gems/parser-3.0.0.0/lib/parser/builders/default.rb:668
       _reduce_102 at jruby/lib/ruby/gems/shared/gems/parser-3.0.0.0/lib/parser/ruby25.rb:3958
  _racc_do_parse_c at com/headius/racc/Cparse.java:709
          do_parse at jruby/lib/ruby/stdlib/racc/parser.rb:264
             parse at jruby/lib/ruby/gems/shared/gems/parser-3.0.0.0/lib/parser/base.rb:190
             parse at jruby/lib/ruby/gems/shared/gems/parser-3.0.0.0/lib/parser/base.rb:33
            <main> at repro.rb:14
```

This bug is probably related to https://github.com/whitequark/parser/issues/305, https://github.com/jruby/jruby/issues/4071, https://github.com/ruby/racc/commit/68bafd39c5df5ea8598fa55964b93f6a123e56b5 and https://github.com/ruby/racc/issues/91, via https://github.com/jruby/jruby/pull/6194.

However my takeaway from those issues and that commit is that this should not be able to happen, as current JRuby should be using racc-1.5+ since at least JRuby 9.1.12.0, which doesn't use RubyContinuation any more. Since it does happen, I wonder whether an old version of the racc cparse.java, possibly in compiled form, lingers in the JRuby codebase and is actually being used.
    
## Further observations
* Although RubyContinuation isn't being used any more, it's still imported into https://github.com/ruby/racc/blob/092af4e2829b9e21e55be321aaa6ce29edf2615c/ext/racc/com/headius/racc/Cparse.java.

* Some racc tests are being skipped on JRuby: https://github.com/ruby/racc/commit/cf377138958e70e1ee2d062abbe9b3c7b7a57ec2, https://github.com/ruby/racc/commit/31072112c4fb158dfda081455afedfde66f3fe1d 
        
**Environment Information**
```
$ bin/jruby -v
jruby 9.2.15.0-SNAPSHOT (2.5.7) 2020-12-25 69334b7b49 OpenJDK 64-Bit Server VM 11.0.9.1+1-Ubuntu-0ubuntu1.20.04 on 11.0.9.1+1-Ubuntu-0ubuntu1.20.04 +jit [linux-x86_64]

$ uname -a
Linux photon 5.4.0-59-generic #65-Ubuntu SMP Thu Dec 10 12:01:51 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by Confusion** [comment]
So this comment is still true: https://github.com/whitequark/parser/issues/305#issuecomment-236805740. 

I think the use of RubyContinuation in the `Cparse.java` was a red herring. The exception is still thrown from `org.jruby.RubyKernel#rbThrowInternal:1232`, while handling the Ruby `throw :racc_jump` on `lib/racc/parser.rb:551`. This looks like it is working as designed (`context.getActiveCatch(tag)` returns a continuation and that continuation is thrown), except that the continuation isn't caught and handled anywhere and is eventually handled by the generic catch block in `org.jruby.ir.interpreter.StartupInterpreterEngine#interpret:109`.

**Comment by Confusion** [comment]
OK, so the answer is actually trivial: the racc on the 9.2 branch is an old version that doesn't include the 1.4.15+ patch to remove the use of `RubyContinuation`. HEAD doesn't have the problem, because it has an upgraded racc.

Is porting the racc update to the 9.2 branch an option?

**Comment by headius** [comment]
@Confusion Are you able to just install the newer racc gem for the `java` platform? That may be all you need to get this working again, assuming we pick up the newer logic.

**Comment by headius** [comment]
I see that we are installing racc as a default gem, but also that we are installing 1.5.1 here: https://github.com/jruby/jruby/blob/9a20077027580422fb720e05d575c5d1039ee829/lib/pom.rb#L38

So it seems we should be installing a pretty recent version, but you say we are not?

**Comment by Confusion** [comment]
> @Confusion Are you able to just install the newer racc gem for the `java` platform? That may be all you need to get this working again, assuming we pick up the newer logic.

Sort of. When using bundler and explicitly requiring racc in the script, that works. Without using bundler or when not explicitly requiring racc, it doesn't. 

> I see that we are installing racc as a default gem, but also that we are installing 1.5.1 here:
> 
> https://github.com/jruby/jruby/blob/9a20077027580422fb720e05d575c5d1039ee829/lib/pom.rb#L38
> 
> So it seems we should be installing a pretty recent version, but you say we are not?

We are, but not yet at the time of the 9.2.14.0 release. I now note that a few things concerning racc have changed since then. E.g. racc sources were removed from the ruby/stdlib directory and racc was added to the default gems, which you linked to. One of those changes solved this issue, because you cannot reproduce it using HEAD. You can reproduce it using the 9.2.14.0 release commit.

So I guess we can consider this solved :shrug:

**Comment by headius** [comment]
> Without using bundler or when not explicitly requiring racc, it doesn't.

I am a bit confused by this... how do you use racc without requiring it?

The explicit require case is below and clearly works ok. Can you show me a session that demonstrates that we are picking up the old version when we should pick up the new version?

```
$ jruby -e "require 'racc'; p Racc::VERSION"
"1.5.1"

$ gem install racc
Fetching racc-1.5.2-java.gem
Successfully installed racc-1.5.2-java
Parsing documentation for racc-1.5.2-java
Installing ri documentation for racc-1.5.2-java
Done installing documentation for racc after 5 seconds
1 gem installed

$ jruby -e "require 'racc'; p Racc::VERSION"
"1.5.2"
```

**Comment by headius** [comment]
To be clear, I am glad this is fixed for 9.2.15 and master, but I would like to understand why the upgraded gem did not get picked up.

**Comment by Confusion** [comment]
I'm not requiring racc myself, because `parser` requires it. 

With jruby 9.2.14.0 or earlier and only the `parser` gem installed:
```
$ jruby -v
jruby 9.2.13.0 (2.5.7) 2020-08-03 9a89c94bcc OpenJDK 64-Bit Server VM 11.0.9.1+1-Ubuntu-0ubuntu1.20.04 on 11.0.9.1+1-Ubuntu-0ubuntu1.20.04 +jit [linux-x86_64]

$ jruby -e "require 'racc'; puts Racc::Parser::Racc_Runtime_Core_Version"
LoadError: no such file to load -- racc

$ jruby -e "require 'parser'; puts Racc::Parser::Racc_Runtime_Core_Version"
1.4.14
```
I putsed that, because Racc::VERSION isn't defined.

**Comment by headius** [comment]
Aha, so this is probably an issue with how the generated parser in `parser` loads the bits of racc that it needs, and those bits are not picking up any installed gem.

I will poke at this a little, since I would like to have closure that the racc gem can be updated properly.

**Comment by headius** [comment]
Ok egg on my face. JRuby master has moved to using racc from the gem, but 9.2 never has received that update. It bundles a copy of the minimal files necessary to support parsers generated from racc, like `parser`, but unless you explicitly activate an installed `racc` gem those copied files will not be updated.

I will propose that we start using `racc` gem in 9.2 as well, even though we may not do a 9.2.15.0 release.

**Comment by headius** [comment]
I have pushed #6517 to switch to using the gem. I don't see any risk here because we already run and maintain fairly recent racc ext in our repository, and this just makes sure the library can be updated independent of JRuby. It also pulls in the non-GPL parts of racc for **generating** parsers, but I think that is fine.

@enebo Any concerns with moving to the racc gem on 9.2 branch?

**Comment by enebo** [comment]
@headius I am ok with moving it.

**Comment by headius** [comment]
This has been resolved via #6517 and I will update master to latest racc so it is in sync.

**Comment by headius** [comment]
Workaround for others that run into this before we release 9.3 or 9.2.15:

* Install the gem and activate it (using `require 'racc'` or `gem 'racc'`) before loading racc or other libraries that depend on it (like parser).
* Install the gem and copy its .jar to stdlib over the one shipped in 9.2.14 and earlier.

**Comment by headius** [comment]
Due to some unforeseen build and deploy issues caused by #6537, we are stepping back and just updating the copied racc parser support files instead of fully making it a default gem.

This issue remains fixed, since racc has been updated, but if future changes require using the racc gem the [workaround I described above](https://github.com/jruby/jruby/issues/6507#issuecomment-754883653) will be required (explicitly activate the racc gem).
